File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0150.cpp
class Solution {
public:
    ///----------------------------------------------------------
    /// Evaluate using a stack
    ///
    int evalRPN(vector<string>& tokens) {
        stack<int> S;

        for (auto token: tokens){
            if (token.length() == 1 && !isdigit(token[0])){
                // If this is an operator, then pop two operands
                // off the stack and apply the operator
                int operand2 = S.top(); S.pop();
                int operand1 = S.top(); S.pop();

                switch(token[0]){
                    case '+': S.push(operand1 + operand2);
                        break;
                    case '-': S.push(operand1 - operand2);
                        break;
                    case '*': S.push(operand1 * operand2);
                        break;
                    case '/': S.push(operand1 / operand2);
                        break;
                } //end-switch

            } else {
                S.push(std::stoi(token));  // Convert the number in string to an integer
            } //end-else
        } //end-for

        return S.top();
    } // end-evalRPN
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0150.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        S = []
        for token in tokens:
            if len(token)==1 and not token.isdigit():
                operand2 = S.pop()
                operand1 = S.pop()
                                
                if token == "+":
                    S.append(operand1 + operand2) 
                elif token == "-":
                    S.append(operand1 - operand2) 
                elif token == "*":
                    S.append(operand1 * operand2) 
                else: # token == "/"
                    S.append(int(operand1 / operand2))
            else:
                S.append(int(token))
        return S[0]
